File: Ahorn/entities/BounceJellyFish.jl
module BounceHelperBounceJellyfish

using ..Ahorn, Maple

@mapdef Entity "BounceHelper/BounceJellyfish" BounceJellyfish(x::Integer, 
                                                              y::Integer,
                                                              platform::Bool=true,
                                                              soulBound::Bool=true,
                                                              baseDashCount::Integer=1,
                                                              ezelMode::Bool=false,
                                                              matchPlayerDash::Bool=false) 

const placements = Ahorn.PlacementDict(
    "Bounce Jellyfish (Bounce Helper)" => Ahorn.EntityPlacement(
        BounceJellyfish
    )
)

const sprites = Dict{Integer, String}(
    0 => "blue",
	1 => "red",
	2 => "pink",
)

function Ahorn.selection(entity::BounceJellyfish)
    sprite = "objects/BounceHelper/bounceJellyfish/blue/idle0"

    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::BounceJellyfish, room::Maple.Room)
    baseDashCount = Int(get(entity.data, "baseDashCount", 1));
    suffix = (get(entity, "ezelMode", false) && baseDashCount > 0) ? "Head/idle0" : "/idle0"; 
    sprite = "objects/BounceHelper/bounceJellyfish/" * sprites[baseDashCount] * suffix;

    Ahorn.drawSprite(ctx, sprite, 0, 0)
    
    if get(entity, "platform", true)
        curve = Ahorn.SimpleCurve((-7, -1), (7, -1), (0, -6))
        Ahorn.drawSimpleCurve(ctx, curve, (1.0, 1.0, 1.0, 1.0), thickness=1)
    end
end

end

File: Loenn/entities/bounceJellyFish.lua
local drawableSprite = require("structs.drawable_sprite")
local drawableLine = require("structs.drawable_line")
local drawing = require("utils.drawing")

local BounceHelperBounceJellyfish = {}

BounceHelperBounceJellyfish.name = "BounceHelper/BounceJellyfish"
BounceHelperBounceJellyfish.depth = -5
local BounceHelperBounceJellyfishDashCounts = {
    0, 1, 2
}

BounceHelperBounceJellyfish.fieldInformation = {
    baseDashCount = {
        options = BounceHelperBounceJellyfishDashCounts,
        editable = false
    }
}
BounceHelperBounceJellyfish.placements = {
    {
        name = "normal",
        data = {
            platform = true,
            soulBound = false,
            baseDashCount = 0,
            ezelMode = false,
            matchPlayerDash = false
        }
    },
    {
        name = "single",
        data = {
            platform = true,
            soulBound = false,
            baseDashCount = 1,
            ezelMode = false,
            matchPlayerDash = false
        }
    },
    {
        name = "double",
        data = {
            platform = true,
            soulBound = false,
            baseDashCount = 2,
            ezelMode = false,
            matchPlayerDash = false
        }
    }
}

local function getColor(entity)
    local dash = entity.baseDashCount

    if dash == 0 then
        return "blue"

    elseif dash == 1 then
        return "red"

    else
        return "pink"
    end
end

function BounceHelperBounceJellyfish.sprite(room, entity)
    local suffix = (entity.ezelMode and entity.baseDashCount > 0) and "Head/idle0" or "/idle0"
    local texture = "objects/BounceHelper/bounceJellyfish/" .. getColor(entity) .. suffix
    local jellySprite = drawableSprite.fromTexture(texture, entity)
    local platform = entity.platform

    if entity.platform then
        local x, y = entity.x or 0, entity.y or 0
        local points = drawing.getSimpleCurve({x - 11, y - 1}, {x + 11, y - 1}, {x - 0, y - 6})
        local lineSprites = drawableLine.fromPoints(points):getDrawableSprite()

        table.insert(lineSprites, 1, jellySprite)

        return lineSprites
    else
        return jellySprite
    end
end

function BounceHelperBounceJellyfish.rectangle(room, entity)
    local texture = "objects/glider/idle0"
    local sprite = drawableSprite.fromTexture(texture, entity)

    return sprite:getRectangle()
end

return BounceHelperBounceJellyfish